File: examples/javascript/advanced-client.js
#!/usr/bin/env node

/**
 * Advanced HTTP Client Example for Knowledge Graph MCP Server
 *
 * This example demonstrates advanced features including:
 * - Comprehensive error handling
 * - Retry logic with exponential backoff
 * - Rate limiting handling
 * - Connection pooling
 * - Batch processing
 * - Performance monitoring
 *
 * Run with: node advanced-client.js
 */

class KnowledgeGraphClient {
	constructor(options = {}) {
		this.baseUrl = options.baseUrl || "http://localhost:3000/api";
		this.timeout = options.timeout || 30000;
		this.maxRetries = options.maxRetries || 3;
		this.retryDelay = options.retryDelay || 1000;
		this.apiKey = options.apiKey;

		// Performance tracking
		this.stats = {
			requests: 0,
			errors: 0,
			totalTime: 0,
			retries: 0,
		};
	}

	async makeRequest(endpoint, options = {}, retryCount = 0) {
		const startTime = Date.now();
		this.stats.requests++;

		const url = `${this.baseUrl}${endpoint}`;
		const requestOptions = {
			timeout: this.timeout,
			headers: {
				"Content-Type": "application/json",
				"X-MCP-Version": "2024-11-05",
				"User-Agent": "KnowledgeGraph-Advanced-Client/1.0.0",
				...(this.apiKey && { Authorization: `Bearer ${this.apiKey}` }),
				...options.headers,
			},
			...options,
		};

		try {
			console.log(`📡 ${options.method || "GET"} ${endpoint}`);

			const response = await fetch(url, requestOptions);
			const data = await response.json();

			// Track timing
			const duration = Date.now() - startTime;
			this.stats.totalTime += duration;

			// Handle rate limiting
			if (response.status === 429) {
				const retryAfter = parseInt(
					response.headers.get("Retry-After") || "60",
				);
				console.log(`⏳ Rate limited. Waiting ${retryAfter} seconds...`);
				await this.sleep(retryAfter * 1000);
				return this.makeRequest(endpoint, options, retryCount);
			}

			// Handle other HTTP errors
			if (!response.ok) {
				const error = new Error(
					`HTTP ${response.status}: ${data.error?.message || "Request failed"}`,
				);
				error.status = response.status;
				error.code = data.error?.code;
				error.details = data.error?.details;
				throw error;
			}

			// Handle API errors
			if (!data.success) {
				const error = new Error(
					`API Error: ${data.error?.message || "Unknown error"}`,
				);
				error.code = data.error?.code;
				error.details = data.error?.details;
				throw error;
			}

			console.log(`✅ Request completed in ${duration}ms`);
			return data.data;
		} catch (error) {
			this.stats.errors++;
			const duration = Date.now() - startTime;
			this.stats.totalTime += duration;

			// Retry logic for certain errors
			if (retryCount < this.maxRetries && this.shouldRetry(error)) {
				this.stats.retries++;
				const delay = this.retryDelay * Math.pow(2, retryCount); // Exponential backoff
				console.log(
					`🔄 Retry ${retryCount + 1}/${this.maxRetries} after ${delay}ms: ${error.message}`,
				);
				await this.sleep(delay);
				return this.makeRequest(endpoint, options, retryCount + 1);
			}

			console.log(`❌ Request failed after ${duration}ms: ${error.message}`);
			throw error;
		}
	}

	shouldRetry(error) {
		// Retry on network errors, timeouts, and 5xx errors (but not 4xx)
		return (
			error.code === "ENOTFOUND" ||
			error.code === "ECONNRESET" ||
			error.code === "ETIMEDOUT" ||
			(error.status >= 500 && error.status < 600)
		);
	}

	async sleep(ms) {
		return new Promise((resolve) => setTimeout(resolve, ms));
	}

	async healthCheck() {
		return this.makeRequest("/health");
	}

	async processKnowledge(text, source, options = {}) {
		return this.makeRequest("/process-knowledge", {
			method: "POST",
			body: JSON.stringify({
				text,
				source,
				thread_id: options.threadId,
				conversation_date: options.conversationDate || new Date().toISOString(),
				include_concepts: options.includeConcepts !== false,
				deduplicate: options.deduplicate !== false,
				processing_batch_id: options.batchId,
			}),
		});
	}

	async searchKnowledge(query, options = {}) {
		return this.makeRequest("/search-knowledge", {
			method: "POST",
			body: JSON.stringify({
				query,
				limit: options.limit || 10,
				threshold: options.threshold || 0.7,
				types: options.types,
				sources: options.sources,
			}),
		});
	}

	async searchConcepts(query, options = {}) {
		return this.makeRequest("/search-concepts", {
			method: "POST",
			body: JSON.stringify({
				query,
				limit: options.limit || 10,
				threshold: options.threshold || 0.7,
			}),
		});
	}

	async storeTriples(triples) {
		return this.makeRequest("/store-triples", {
			method: "POST",
			body: JSON.stringify({ triples }),
		});
	}

	async deduplicateTriples(triples) {
		return this.makeRequest("/deduplicate", {
			method: "POST",
			body: JSON.stringify({ triples }),
		});
	}

	async getStats() {
		return this.makeRequest("/stats");
	}

	async getEntities(options = {}) {
		const params = new URLSearchParams();
		if (options.role) params.set("role", options.role);
		if (options.minOccurrence)
			params.set("min_occurrence", options.minOccurrence);
		if (options.limit) params.set("limit", options.limit);
		if (options.sortBy) params.set("sort_by", options.sortBy);
		if (options.sources) params.set("sources", options.sources.join(","));
		if (options.types) params.set("types", options.types.join(","));

		const query = params.toString();
		return this.makeRequest(`/entities${query ? "?" + query : ""}`);
	}

	async getVersion() {
		return this.makeRequest("/version");
	}

	async getMetrics() {
		return this.makeRequest("/metrics");
	}

	async getCapabilities() {
		return this.makeRequest("/capabilities");
	}

	// Batch processing with concurrency control
	async processBatch(items, processor, options = {}) {
		const concurrency = options.concurrency || 3;
		const results = [];
		const errors = [];

		console.log(
			`📦 Processing batch of ${items.length} items with concurrency ${concurrency}`,
		);

		for (let i = 0; i < items.length; i += concurrency) {
			const chunk = items.slice(i, i + concurrency);
			const promises = chunk.map(async (item, index) => {
				try {
					const result = await processor(item, i + index);
					return { success: true, index: i + index, result };
				} catch (error) {
					return { success: false, index: i + index, error: error.message };
				}
			});

			const chunkResults = await Promise.all(promises);

			for (const result of chunkResults) {
				if (result.success) {
					results.push(result);
				} else {
					errors.push(result);
					console.log(`❌ Batch item ${result.index} failed: ${result.error}`);
				}
			}

			// Progress update
			console.log(
				`📊 Progress: ${Math.min(i + concurrency, items.length)}/${items.length} items processed`,
			);

			// Rate limiting pause between chunks
			if (i + concurrency < items.length) {
				await this.sleep(100);
			}
		}

		console.log(
			`✅ Batch completed: ${results.length} succeeded, ${errors.length} failed`,
		);
		return { results, errors };
	}

	getPerformanceStats() {
		const avgResponseTime =
			this.stats.requests > 0 ? this.stats.totalTime / this.stats.requests : 0;
		const errorRate =
			this.stats.requests > 0
				? (this.stats.errors / this.stats.requests) * 100
				: 0;

		return {
			totalRequests: this.stats.requests,
			totalErrors: this.stats.errors,
			totalRetries: this.stats.retries,
			averageResponseTime: Math.round(avgResponseTime),
			errorRate: Math.round(errorRate * 100) / 100,
			totalTime: this.stats.totalTime,
		};
	}

	resetStats() {
		this.stats = {
			requests: 0,
			errors: 0,
			totalTime: 0,
			retries: 0,
		};
	}
}

async function main() {
	const client = new KnowledgeGraphClient({
		baseUrl: "http://localhost:3000/api",
		timeout: 30000,
		maxRetries: 3,
		retryDelay: 1000,
	});

	try {
		console.log("🚀 Knowledge Graph MCP Server - Advanced Client Example\n");

		// 1. Health check
		console.log("1. Performing health check...");
		const health = await client.healthCheck();
		console.log("✅ Server health:", health.status);
		console.log();

		// 2. Get server information
		console.log("2. Getting server information...");
		const [version, capabilities, metrics] = await Promise.all([
			client.getVersion(),
			client.getCapabilities(),
			client.getMetrics(),
		]);
		console.log("✅ Server version:", version.version);
		console.log(
			"✅ Available tools:",
			capabilities.tools.map((t) => t.name).join(", "),
		);
		console.log("✅ Server metrics:", {
			uptime: metrics.uptime,
			memoryUsage: `${Math.round(metrics.memoryUsage.heapUsed / 1024 / 1024)}MB`,
		});
		console.log();

		// 3. Batch processing example
		console.log("3. Demonstrating batch processing...");
		const texts = [
			"John is a data scientist at Google. He loves Python and machine learning.",
			"Sarah works at Microsoft as a software engineer. She specializes in cloud computing.",
			"Alex is a product manager at Apple. He focuses on user experience design.",
			"Maria is a researcher at Stanford University. She studies artificial intelligence.",
			"David is a DevOps engineer at Amazon. He works with containerization and Kubernetes.",
		];

		const batchResult = await client.processBatch(
			texts,
			async (text, index) => {
				return await client.processKnowledge(text, `batch_example_${index}`, {
					threadId: "advanced_demo",
					includeConcepts: true,
				});
			},
			{ concurrency: 2 },
		);

		console.log(
			`✅ Batch processing completed: ${batchResult.results.length} items processed`,
		);
		console.log();

		// 4. Advanced search with multiple criteria
		console.log("4. Performing advanced search...");
		const searchResults = await client.searchKnowledge(
			"software engineer data scientist machine learning",
			{
				limit: 10,
				threshold: 0.6,
				types: ["entity-entity", "entity-event"],
				sources: ["batch_example_0", "batch_example_1", "batch_example_3"],
			},
		);

		console.log(
			`✅ Advanced search found ${searchResults.results.length} results`,
		);
		searchResults.results.slice(0, 3).forEach((result, index) => {
			console.log(
				`   ${index + 1}. ${result.triple.subject} → ${result.triple.predicate} → ${result.triple.object}`,
			);
			console.log(
				`      Score: ${result.similarity.toFixed(3)} | Type: ${result.triple.type}`,
			);
		});
		console.log();

		// 5. Entity analysis
		console.log("5. Analyzing entities...");
		const entities = await client.getEntities({
			role: "both",
			minOccurrence: 1,
			limit: 20,
			sortBy: "frequency",
		});

		console.log("✅ Top entities by frequency:");
		entities.entities.slice(0, 5).forEach((entity, index) => {
			console.log(
				`   ${index + 1}. "${entity.entity}" (${entity.frequency} occurrences)`,
			);
		});
		console.log();

		// 6. Concept exploration
		console.log("6. Exploring concepts...");
		const concepts = await client.searchConcepts(
			"technology software programming",
			{
				limit: 5,
				threshold: 0.7,
			},
		);

		console.log("✅ Related concepts:");
		concepts.results.forEach((result, index) => {
			console.log(
				`   ${index + 1}. ${result.concept.concept} (${result.concept.abstraction_level})`,
			);
			console.log(`      Similarity: ${result.similarity.toFixed(3)}`);
		});
		console.log();

		// 7. Performance statistics
		console.log("7. Performance statistics:");
		const stats = client.getPerformanceStats();
		console.log("✅ Client performance:");
		console.log(`   Total requests: ${stats.totalRequests}`);
		console.log(`   Average response time: ${stats.averageResponseTime}ms`);
		console.log(`   Error rate: ${stats.errorRate}%`);
		console.log(`   Total retries: ${stats.totalRetries}`);
		console.log();

		console.log("🎉 Advanced client example completed successfully!");
	} catch (error) {
		console.error("❌ Critical error:", error.message);
		if (error.details) {
			console.error("📝 Error details:", error.details);
		}
		process.exit(1);
	}
}

// Handle graceful shutdown
process.on("SIGINT", () => {
	console.log("\n🛑 Received SIGINT, shutting down gracefully...");
	process.exit(0);
});

process.on("SIGTERM", () => {
	console.log("\n🛑 Received SIGTERM, shutting down gracefully...");
	process.exit(0);
});

// Handle unhandled promise rejections
process.on("unhandledRejection", (reason, promise) => {
	console.error("Unhandled Rejection at:", promise, "reason:", reason);
	process.exit(1);
});

// Run the example
main();

File: examples/python/advanced_client.py
#!/usr/bin/env python3

"""
Advanced HTTP Client Example for Knowledge Graph MCP Server

This example demonstrates advanced features including:
- Async/await support with aiohttp
- Comprehensive error handling
- Retry logic with exponential backoff
- Rate limiting handling
- Connection pooling
- Batch processing
- Performance monitoring

Requirements:
    pip install aiohttp asyncio tenacity
"""

import asyncio
import json
import time
from datetime import datetime
from typing import Dict, Any, Optional, List, Union
from dataclasses import dataclass, field

import aiohttp
from tenacity import retry, stop_after_attempt, wait_exponential, retry_if_exception_type


@dataclass
class ClientStats:
    """Track client performance statistics"""
    requests: int = 0
    errors: int = 0
    total_time: float = 0.0
    retries: int = 0
    
    @property
    def average_response_time(self) -> float:
        return self.total_time / self.requests if self.requests > 0 else 0.0
    
    @property
    def error_rate(self) -> float:
        return (self.errors / self.requests * 100) if self.requests > 0 else 0.0


class KnowledgeGraphAdvancedClient:
    def __init__(self, 
                 base_url: str = "http://localhost:3000/api",
                 timeout: int = 30,
                 max_retries: int = 3,
                 retry_delay: float = 1.0,
                 api_key: Optional[str] = None,
                 connector_limit: int = 100):
        
        self.base_url = base_url.rstrip('/')
        self.timeout = aiohttp.ClientTimeout(total=timeout)
        self.max_retries = max_retries
        self.retry_delay = retry_delay
        self.api_key = api_key
        
        # Performance tracking
        self.stats = ClientStats()
        
        # Session configuration
        connector = aiohttp.TCPConnector(
            limit=connector_limit,
            limit_per_host=20,
            keepalive_timeout=30,
            enable_cleanup_closed=True
        )
        
        headers = {
            'Content-Type': 'application/json',
            'X-MCP-Version': '2024-11-05',
            'User-Agent': 'KnowledgeGraph-Advanced-Python-Client/1.0.0'
        }
        
        if self.api_key:
            headers['Authorization'] = f'Bearer {self.api_key}'
        
        self.session = aiohttp.ClientSession(
            connector=connector,
            headers=headers,
            timeout=self.timeout
        )
    
    async def __aenter__(self):
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()
    
    async def close(self):
        """Close the HTTP session"""
        if self.session and not self.session.closed:
            await self.session.close()
    
    @retry(
        stop=stop_after_attempt(3),
        wait=wait_exponential(multiplier=1, min=1, max=10),
        retry=retry_if_exception_type((aiohttp.ClientError, asyncio.TimeoutError))
    )
    async def _make_request(self, endpoint: str, method: str = 'GET', data: Optional[Dict] = None) -> Dict[str, Any]:
        """Make an HTTP request with retry logic"""
        start_time = time.time()
        self.stats.requests += 1
        
        url = f"{self.base_url}{endpoint}"
        
        try:
            print(f"📡 {method} {endpoint}")
            
            async with self.session.request(method, url, json=data) as response:
                response_data = await response.json()
                
                # Track timing
                duration = time.time() - start_time
                self.stats.total_time += duration
                
                # Handle rate limiting
                if response.status == 429:
                    retry_after = int(response.headers.get('Retry-After', 60))
                    print(f"⏳ Rate limited. Waiting {retry_after} seconds...")
                    await asyncio.sleep(retry_after)
                    self.stats.retries += 1
                    return await self._make_request(endpoint, method, data)
                
                # Handle HTTP errors
                if response.status >= 400:
                    error_msg = response_data.get('error', {}).get('message', f'HTTP {response.status}')
                    raise aiohttp.ClientResponseError(
                        request_info=response.request_info,
                        history=response.history,
                        status=response.status,
                        message=error_msg
                    )
                
                # Handle API errors
                if not response_data.get('success', True):
                    error_msg = response_data.get('error', {}).get('message', 'Unknown API error')
                    raise RuntimeError(f"API Error: {error_msg}")
                
                print(f"✅ Request completed in {duration*1000:.1f}ms")
                return response_data.get('data', response_data)
                
        except Exception as e:
            self.stats.errors += 1
            duration = time.time() - start_time
            self.stats.total_time += duration
            print(f"❌ Request failed after {duration*1000:.1f}ms: {str(e)}")
            raise
    
    async def health_check(self) -> Dict[str, Any]:
        """Check server health status"""
        return await self._make_request('/health')
    
    async def process_knowledge(self, text: str, source: str, 
                               thread_id: Optional[str] = None,
                               conversation_date: Optional[str] = None,
                               include_concepts: bool = True,
                               deduplicate: bool = True,
                               batch_id: Optional[str] = None) -> Dict[str, Any]:
        """Extract and store knowledge from text"""
        data = {
            'text': text,
            'source': source,
            'include_concepts': include_concepts,
            'deduplicate': deduplicate
        }
        
        if thread_id:
            data['thread_id'] = thread_id
        if conversation_date:
            data['conversation_date'] = conversation_date
        else:
            data['conversation_date'] = datetime.utcnow().isoformat() + 'Z'
        if batch_id:
            data['processing_batch_id'] = batch_id
        
        return await self._make_request('/process-knowledge', 'POST', data)
    
    async def search_knowledge(self, query: str, 
                              limit: int = 10,
                              threshold: float = 0.7,
                              types: Optional[List[str]] = None,
                              sources: Optional[List[str]] = None) -> Dict[str, Any]:
        """Search the knowledge graph"""
        data = {
            'query': query,
            'limit': limit,
            'threshold': threshold
        }
        
        if types:
            data['types'] = types
        if sources:
            data['sources'] = sources
        
        return await self._make_request('/search-knowledge', 'POST', data)
    
    async def search_concepts(self, query: str,
                             limit: int = 10,
                             threshold: float = 0.7) -> Dict[str, Any]:
        """Search concepts in the knowledge graph"""
        data = {
            'query': query,
            'limit': limit,
            'threshold': threshold
        }
        
        return await self._make_request('/search-concepts', 'POST', data)
    
    async def store_triples(self, triples: List[Dict[str, Any]]) -> Dict[str, Any]:
        """Store pre-structured knowledge triples"""
        return await self._make_request('/store-triples', 'POST', {'triples': triples})
    
    async def deduplicate_triples(self, triples: List[Dict[str, Any]]) -> Dict[str, Any]:
        """Deduplicate knowledge triples"""
        return await self._make_request('/deduplicate', 'POST', {'triples': triples})
    
    async def get_stats(self) -> Dict[str, Any]:
        """Get knowledge graph statistics"""
        return await self._make_request('/stats')
    
    async def get_entities(self, role: str = 'both',
                          min_occurrence: int = 1,
                          limit: int = 100,
                          sort_by: str = 'frequency',
                          sources: Optional[List[str]] = None,
                          types: Optional[List[str]] = None) -> Dict[str, Any]:
        """Get entities from the knowledge graph"""
        params = [
            f"role={role}",
            f"min_occurrence={min_occurrence}",
            f"limit={limit}",
            f"sort_by={sort_by}"
        ]
        
        if sources:
            params.append(f"sources={','.join(sources)}")
        if types:
            params.append(f"types={','.join(types)}")
        
        query_string = '?' + '&'.join(params)
        return await self._make_request(f'/entities{query_string}')
    
    async def get_version(self) -> Dict[str, Any]:
        """Get server version information"""
        return await self._make_request('/version')
    
    async def get_metrics(self) -> Dict[str, Any]:
        """Get server metrics"""
        return await self._make_request('/metrics')
    
    async def get_capabilities(self) -> Dict[str, Any]:
        """Get server capabilities"""
        return await self._make_request('/capabilities')
    
    async def process_batch(self, items: List[Any], processor, concurrency: int = 3) -> Dict[str, Any]:
        """Process a batch of items with controlled concurrency"""
        semaphore = asyncio.Semaphore(concurrency)
        results = []
        errors = []
        
        print(f"📦 Processing batch of {len(items)} items with concurrency {concurrency}")
        
        async def process_item(item, index):
            async with semaphore:
                try:
                    result = await processor(item, index)
                    return {'success': True, 'index': index, 'result': result}
                except Exception as e:
                    return {'success': False, 'index': index, 'error': str(e)}
        
        # Process all items
        tasks = [process_item(item, i) for i, item in enumerate(items)]
        batch_results = await asyncio.gather(*tasks, return_exceptions=True)
        
        # Categorize results
        for result in batch_results:
            if isinstance(result, Exception):
                errors.append({'error': str(result)})
            elif result['success']:
                results.append(result)
            else:
                errors.append(result)
                print(f"❌ Batch item {result['index']} failed: {result['error']}")
        
        print(f"✅ Batch completed: {len(results)} succeeded, {len(errors)} failed")
        return {'results': results, 'errors': errors}
    
    def get_performance_stats(self) -> Dict[str, Union[int, float]]:
        """Get client performance statistics"""
        return {
            'total_requests': self.stats.requests,
            'total_errors': self.stats.errors,
            'total_retries': self.stats.retries,
            'average_response_time': round(self.stats.average_response_time * 1000, 1),  # ms
            'error_rate': round(self.stats.error_rate, 2),
            'total_time': round(self.stats.total_time, 2)
        }
    
    def reset_stats(self):
        """Reset performance statistics"""
        self.stats = ClientStats()


async def main():
    async with KnowledgeGraphAdvancedClient(
        base_url='http://localhost:3000/api',
        timeout=30,
        max_retries=3,
        retry_delay=1.0
    ) as client:
        
        try:
            print("🚀 Knowledge Graph MCP Server - Advanced Python Client Example\n")
            
            # 1. Health check and server info
            print("1. Performing health check and getting server info...")
            health, version, capabilities, metrics = await asyncio.gather(
                client.health_check(),
                client.get_version(),
                client.get_capabilities(),
                client.get_metrics()
            )
            
            print(f"✅ Server health: {health['status']}")
            print(f"✅ Server version: {version['version']}")
            print(f"✅ Available tools: {', '.join([tool['name'] for tool in capabilities['tools']])}")
            print(f"✅ Server uptime: {metrics['uptime']}")
            print()
            
            # 2. Batch processing example
            print("2. Demonstrating batch processing...")
            texts = [
                "Rachel is a data scientist at Netflix. She specializes in recommendation algorithms and machine learning.",
                "Tom works as a DevOps engineer at Spotify. He focuses on Kubernetes and infrastructure automation.",
                "Lisa is a product manager at Airbnb. She works on user experience and growth strategies.",
                "James is a security researcher at Google. He specializes in cryptography and secure systems.",
                "Anna is a frontend developer at Facebook. She works with React and modern web technologies."
            ]
            
            async def process_text(text, index):
                return await client.process_knowledge(
                    text, 
                    f'batch_example_{index}',
                    thread_id='advanced_demo',
                    include_concepts=True
                )
            
            batch_result = await client.process_batch(texts, process_text, concurrency=2)
            print(f"✅ Batch processing completed: {len(batch_result['results'])} items processed")
            print()
            
            # 3. Advanced search with multiple criteria
            print("3. Performing advanced search...")
            search_results = await client.search_knowledge(
                'data scientist machine learning DevOps Kubernetes',
                limit=10,
                threshold=0.6,
                types=['entity-entity', 'entity-event'],
                sources=[f'batch_example_{i}' for i in range(5)]
            )
            
            print(f"✅ Advanced search found {len(search_results['results'])} results")
            for i, result in enumerate(search_results['results'][:3], 1):
                triple = result['triple']
                print(f"   {i}. {triple['subject']} → {triple['predicate']} → {triple['object']}")
                print(f"      Score: {result['similarity']:.3f} | Type: {triple['type']}")
            print()
            
            # 4. Concurrent operations
            print("4. Performing concurrent operations...")
            stats_task = client.get_stats()
            entities_task = client.get_entities(role='both', min_occurrence=1, limit=20, sort_by='frequency')
            concepts_task = client.search_concepts('technology software programming', limit=5, threshold=0.7)
            
            stats, entities, concepts = await asyncio.gather(stats_task, entities_task, concepts_task)
            
            print("✅ Concurrent operations completed:")
            print(f"   Knowledge graph stats: {stats['totalTriples']} triples, {stats['totalConcepts']} concepts")
            print(f"   Top entities: {', '.join([e['entity'] for e in entities['entities'][:3]])}")
            print(f"   Related concepts: {', '.join([c['concept']['concept'] for c in concepts['results'][:3]])}")
            print()
            
            # 5. Error handling demonstration
            print("5. Demonstrating error handling...")
            try:
                # This should fail due to invalid parameters
                await client.search_knowledge("", limit=-1, threshold=2.0)
            except Exception as e:
                print(f"✅ Correctly handled invalid request: {type(e).__name__}")
            print()
            
            # 6. Performance statistics
            print("6. Performance statistics:")
            perf_stats = client.get_performance_stats()
            print("✅ Client performance:")
            for key, value in perf_stats.items():
                print(f"   {key.replace('_', ' ').title()}: {value}")
            print()
            
            print("🎉 Advanced Python client example completed successfully!")
            
        except Exception as error:
            print(f"❌ Critical error: {error}")
            
            if "Connection refused" in str(error):
                print("💡 Make sure the Knowledge Graph MCP Server is running with HTTP transport enabled:")
                print("   ENABLE_HTTP_TRANSPORT=true pnpm run dev:http")
            
            raise


if __name__ == "__main__":
    asyncio.run(main())